### ISSUE TITLE
Anonyomus classes in 2.3 (test data provider)

### ISSUE THREAD

**Issue by oskar1233** [magento/magento2#21555]
### Preconditions (*)

1. Magento 2.3
2. PHP 7.2

### Steps to reproduce (*)

1. Create a new module
2. Create a test with data provider returning anonymous class
3. `./bin/magento setup:di:compile`

### Expected result (*)

1. Compilation succed

### Actual result (*)

1. `Notice: Uninitialized string offset: 1 in /var/www/setup/src/Magento/Setup/Module/Di/Code/Scanner/PhpScanner.php on line 219`

My sample test class looks like this:
```
<?php

namespace OskarLaska\Xyz\Test\Unit;

class TestTest extends \PHPUnit\Framework\TestCase
{
    /**
     * @param $theObject
     *
     * @dataProvider someProvider
     */
    public function testX($theObject)
    {
        $this->assertEquals('It works!', $theObject->callMe());
    }

    public function someProvider()
    {
        return [
            'test' => [
                new class {
                    public function callMe() {
                        return 'It works!';
                    }
                }
            ]
        ];
    }
}
```

Not sure if the anonymous class needs to be in data provider of PHPUnit's test. I saw other issues with anonymous classes resolved hence I provide my specific case.

**Comment by magento-engcom-team** [comment]
Hi @oskar1233. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@oskar1233 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by oskar1233** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @oskar1233. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @oskar1233, here is your [Magento instance](https://i-21555-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-21555-2-3-develop.instances.magento-community.engineering/admin](https://i-21555-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @oskar1233 thank you for you report, i;m not able to reproduce following steps you described.
![deepinscreenshot_select-area_20190304103745](https://user-images.githubusercontent.com/40756439/53720538-aad3b480-3e69-11e9-8b69-529199e55689.png)
![deepinscreenshot_select-area_20190304103754](https://user-images.githubusercontent.com/40756439/53720539-ab6c4b00-3e69-11e9-9d95-ff142fa5bf3e.png)

**Comment by oskar1233** [comment]
@engcom-backlog-nazar I use PHP 7.2 to be clear - forgot to mention, sorry.

I'm pretty confident the bug exists; it appeared on my project and then I was able to reproduce it.

**Comment by ghost** [comment]
@oskar1233 ok i'm will recheck on php 7.2

**Comment by oskar1233** [comment]
@engcom-backlog-nazar Also, I have created a module using separate composer project. Sorry for not mentioning, I thought it's a preferred way to do so.

**Comment by oskar1233** [comment]
@engcom-backlog-nazar Any progress on this one?

**Comment by ghost** [comment]
@oskar1233 same result with php7.2 can you share with your test module ?

**Comment by slavvka** [comment]
Hi @oskar1233. Thank you for your report. 
The issue has been fixed in magento/magento2#26533 by @joni-jones in 2.4-develop branch
Related commit(s):
 - [fd3f2a22b8e5b2cffaf17daccc29f461c26c5f51](https://github.com/magento/magento2/commit/fd3f2a22b8e5b2cffaf17daccc29f461c26c5f51)
 - [136d5592244d05509ad0b3b9117505d8f833ddfa](https://github.com/magento/magento2/commit/136d5592244d05509ad0b3b9117505d8f833ddfa)
 - [25fd9259e72bc484f5899aa4edf3de120b66db0c](https://github.com/magento/magento2/commit/25fd9259e72bc484f5899aa4edf3de120b66db0c)
 - [e165d99f5994af39d49ed70155c003745158e5db](https://github.com/magento/magento2/commit/e165d99f5994af39d49ed70155c003745158e5db)

The fix will be available with the upcoming 2.4.0 release.
